### DISCUSSION TITLE
Annotations + annotation pages

### DISCUSSION THREAD

**Discussion by stephenwf** [digirati-co-uk/iiif-canvas-panel#99] [Ideas]
In order to provide a way to display one or more annotations on a given canvas, we need to have a solid model for:

* Discovering available annotation pages
* Visibility of annotation pages (Enabling/disabling) 
* Styling/customising annotation pages (to propagate down to annotations)
* Styling/customising individual annotations (e.g. to drive selection state)
* Normalisation of targets/selectors/bodies
* Loading external annotation lists

In addition to this, it should be possible to group the visibility together. This allows for switching "views", for example if you had a view set up with N annotation pages and then performed a search query, you could swap the "view" to be driven by the search results. Once the search is complete, you can flip back to your previous "view".

In terms of rendering, the annotations that need to be rendered are part of the Rendering strategy (#83). Only the identifiers are returned here, and deducing the styling/customisations/normalisation is done per-annotation.

### Discovering available annotation pages

There are various sources for available annotation pages when viewing a canvas:
* On the manifest, inline annotations property **(TBC)**
* On the manifest, dereferencable annotations (items == undefined) **(TBC)**
* On the canvas, inline annotation property
* On the canvas, dereferencable annotation property (items == undefined)
* On a range `supplementary` property **(TBC)**
* On a range `annotations` property **(TBC)**
* Custom annotation page loaded outside of the manifest (associated in the store, see below)

This will be a mix of loaded and unloaded annotation lists. Some will have labels, some will not have labels. The UI rendering must decide which to enable, and how to render a UI to enable one or more.

### Visibility of annotation pages

When you get a list of available annotation pages, you will also get a functional API similar to this:
```js
setPageEnabled(annotationPageId, { deselectOthers: false });
setPageDisabled(annotationPageId);
loadAnnotationPage(annotationPageId);
```

They are stored in the Vault meta (TODO: Link to docs) for each annotation page:
```js
const state = {
  meta: {
    'https://example.org/annotation-page/1': {
      annotationPageManager: {
        isActive: true,
        isLoading: false,
        isLoaded: false,
        didError: false,
        views: ['https://example.org/manifest-1'],
      }
    }
  }
}
```

When you request "visible" annotation pages, you choose a "view". By default this is the manifest ID. 

### Styling annotations
Although styling of individual annotations has been discussed in the documentation, it may be better for developers to style entire pages of annotations, so any new annotations that get added inherit that style, and leave annotation styles as overrides. This would be another entry in the `meta` for that annotation pages. **These would not be scoped to views, like the above**

Example `meta` entry for an annotation page (Attributes not final)
```js
const state = {
  meta: {
    'https://example.org/annotation-page/1': {
      annotationStyles: {
        background: 'rgba(0, 0, 0, 0.5)',
        borderWidth: '1px',
        viewerSizing: 'viewer', // 1px = 1px on the viewport, not on the canvas
      }
    }
  }
}
```

The above would also be valid on annotations themselves. These properties would be merged when you choose a rendering strategy for annotations. It would then be up to the renderer to use this style information to render the annotations. Atlas will aim to support as much as possible both for HTML-based annotations and also on-canvas annotations (where possible). 

### Normalising targets, selectors and bodies.

This is typically where annotations get complicated. This will be covered in another discussion. The short is that targets, selectors and bodies will first (on the fly, not in Vault) be expanded to their most verbose form to de-duplicate similar ways of writing the same thing, and then reframed into a set of known targets, selectors and bodies. These well-knowns can then be used as a check list for rendering (on viewer and outside of viewer).

### How this relates to search 
Search is like parameterised annotation pages (wrapped in collections).This could be the conceptual model we use for it in Vault. This would also open the door for custom parameterisations. We have the model here defined for "visible" parameter attached to pages, but you could see a custom implementation - offline text search on loaded annotations for example - could use a similar mechanism.

## Questions

* [ ] The spec says that annotations must target the resource they are attached to, but could they deference something inside the resource too? For example, could a `manifest.annotations`  contain annotations targeting canvases? (same question for ranges)
* [ ] What should the default "view" for annotation pages be? Manifest is chosen as it seems reasonable that a user may create a new annotation page targeting multiple canvases to drive a narrative or similar.
* [ ] When advertising annotation lists - [how do we account for 2-up views or similar?](https://github.com/digirati-co-uk/iiif-canvas-panel/discussions/100)

## Development tasks

* [ ] Change `resources` to `views` in page manager
* [ ] Add loading functions and tracking loading state
* [ ] Visible way in the UI to react to loading status
* [ ] Add new "VisibleCanvases" context and hook, use this for advertising annotation pages (and possibly split rendering and advertising)

**Comment by tomcrane** [comment]
>  For example, could a manifest.annotations contain annotations targeting canvases?

I think the answer to this is yes, as canvases are part of the manifest. It would not be good practice to put canvas-level tags or transcribed blocks in manifest.annotations, but it _would_ be a sensible place for annos targeting multiple canvases, or convenience lists of certain annos from across the manifest, or for specialised clients like a narrative/slideshow/exhibition where the manifest is the subject and the single anno list is the guided viewing experience.

**Comment by tomcrane** [comment]
I'm interested in where the management of annos in the Vault crosses over to the visual display and affordances of annotations on a canvas panel instance.

The doc site currently distinguishes between the anno in Vault, which is a normalised W3C anno that vault has loaded and is tracking for you (a piece of data), with the visible and potentially interactive manifestation of an annotation on the canvas panel surface - an anno that's also now part of the DOM, if you see what I mean - something that can have borders, colours, draggability, clickability, raise events from interactions and so on... is part of the UI.

Vault might have 1000s of annos loaded; perhaps you have only followed anno page links to get the annos for the current canvas, you have _data_ in Vault for those annos but they become something else again to be rendered as pixels on screen.

That `DisplayAnnotation` in the docs is purely a Canvas Panel concept, it wouldn't mean anything in Vault, it's the bridge between user actions to view and even modify the anno and the data in Vault. A user-initiated change to a DisplayAnnotation might update the Vault W3C anno's w,h properties, for example. `DisplayAnnotation` is the interface for an Anno as an interactive visual thing in the browser.

In the _Styling Annotations_ section - is that a way to turn W3C Vault Annos into CP DisplayAnnotations in bulk (bulk styling)? And `setPageEnabled()` etc... could that switch the set of CP-level DisplayAnnotations as interactive elements, from the W3C Web Annos?

Or is this different from the current docs ... are all W3C annos in the vault _also_ styleable interactive UI elements (at least, potentially)?

**Reply by tomcrane** [reply]
Doc link - https://deploy-preview-50--iiif-canvas-panel.netlify.app/docs/examples/annotations#displayannotation
